Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp

```diff
@@ -247,6 +267,80 @@ String XMLHttpRequest::get_text_response() const
+    if (!m_response->body())
+        return;
+
+    // 2. Let finalMIME be the result of get a final MIME type for xhr.
+    auto final_mine = MUST(get_final_mime_type());
+
+    // 3. If finalMIME is not an HTML MIME type or an XML MIME type, then return.
+    if (!final_mine.is_html() && !final_mine.is_xml())
+        return;
+
+    // 4. If xhr’s response type is the empty string and finalMIME is an HTML MIME type, then return.
+    if (m_response_type == Bindings::XMLHttpRequestResponseType::Empty && final_mine.is_html())
+        return;
+
+    // 5. If finalMIME is an HTML MIME type, then:
+    Optional<StringView> charset;
+    JS::GCPtr<DOM::Document> document;
+    if (final_mine.is_html()) {
+        // 5.1. Let charset be the result of get a final encoding for xhr.
+        charset = MUST(get_final_encoding());
+
+        // 5.2. If charset is null, prescan the first 1024 bytes of xhr’s received bytes and if that does not terminate unsuccessfully then let charset be the return value.
+        document = DOM::Document::create(realm());
+        if (!charset.has_value())
+            if (auto found_charset = HTML::run_prescan_byte_stream_algorithm(*document, m_received_bytes); found_charset.has_value())
+                charset = MUST(String::from_deprecated_string(found_charset.value())).bytes_as_string_view();
+
+        // 5.3. If charset is null, then set charset to UTF-8.
+        if (!charset.has_value())
+            charset = "UTF-8"sv;
+
+        // 5.4. Let document be a document that represents the result parsing xhr’s received bytes following the rules set forth in the HTML Standard for an HTML parser with scripting disabled and a known definite encoding charset.
+        auto parser = HTML::HTMLParser::create(*document, m_received_bytes, charset.value());
+        parser->run(document->url());
+
+        // 5.5. Flag document as an HTML document.
+        document->set_document_type(DOM::Document::Type::HTML);
+    }
+
+    // 6. Otherwise, let document be a document that represents the result of running the XML parser with XML scripting support disabled on xhr’s received bytes. If that fails (unsupported character encoding, namespace well-formedness error, etc.), then return null.
+    else {
+        document = DOM::Document::create(realm());
+        if (!Web::build_xml_document(*document, m_received_bytes)) {
+            m_response_object = JS::js_null();
+            m_response_document = document;
+            return;
+        }
+    }
+
+    // 7. If charset is null, then set charset to UTF-8.
+    if (!charset.has_value())
+        charset = "UTF-8"sv;
+
+    // 8. Set document’s encoding to charset.
+    document->set_encoding(charset->to_deprecated_string());
+
+    // 9. Set document’s content type to finalMIME.
+    document->set_content_type(MUST(final_mine.serialized()).to_deprecated_string());
+
+    // 10. Set document’s URL to xhr’s response’s URL.
+    document->set_url(m_request_url);
```

Comment:
I may be mistaken (trying to get comfortable with reading specs let alone the XHR spec!) but I believe we should be passing response's URL instead of the request URL here, something like this I would think:

```cpp
document->set_url(m_response->url().value_or({});
```
Though from a quick read of the Fetch spec, it seems that a [response's URL](https://fetch.spec.whatwg.org/#concept-response-url) can be a `URL` or null which is why I include the `value_or()` call. However, I'm not sure if it makes sense to set the document URL to an empty `URL`.

Reply:
You are right I missed that one